### PR TITLE
Added native support for the HTTP PATCH verb.

### LINKED ISSUE #17299: HttpMethod should include "PATCH" as a property
# Latest Proposal

Expose Http `PATCH` Verb alongside [existing verbs on `HttpMethod`](https://apisof.net/catalog/System.Net.Http.HttpMethod) `GET`, `PUT`, `POST`, `DELETE`, etc. and on [`HttpClient` as Async methods](https://apisof.net/catalog/System.Net.Http.HttpClient)

```diff
 namespace System.Net.Http {
     public class HttpClient {
+        public Task<HttpResponseMessage> PatchAsync(string requestUri, HttpContent content);
+        public Task<HttpResponseMessage> PatchAsync(Uri requestUri, HttpContent content);
+        public Task<HttpResponseMessage> PatchAsync(string requestUri, HttpContent content, CancellationToken cancellationToken);
+        public Task<HttpResponseMessage> PatchAsync(Uri requestUri, HttpContent content, CancellationToken cancellationToken);
     }
     public class HttpMethod {
+        public static HttpMethod Patch { get; }
     }
 }
```

# Original Proposal

## Description

As a part of Http requests, PATCH is an often used way to do a partial update of an object, and is common enough to merit it's own property within the `HttpMethod` class.

## Proposed API Changes

```csharp
HttpMethod.Patch;
```

## Example

```csharp
// old
var request = new HttpRequestMessage
{
    Method = new HttpMethod("PATCH") // deviates from Get, Put, Post, Delete
};

// new
var request = new HttpRequestMessage
{
    Method = HttpMethod.Patch // matches Get, Put, Post, Delete
};
```

I'd like to submit a PR with the simple change to add PATCH as an Http Method in order to round out the class. This change in the future could also extend the `HttpClient` to include new `PatchAsync` methods.

### PR DESCRIPTION
**Pull request by TylerBrinkley**
Added native support for the HTTP PATCH verb.

I didn't see any existing tests for the other verbs equivalent methods and properties so no unit tests are included.

Closes #17299

### CODE REVIEW

**Comment by davidsh** [comment]
Adding new APIs require modifying the REF folder as well:

https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/ref/System.Net.Http.cs

**Comment by TylerBrinkley** [comment]
@davdish Thanks, done.

**Comment by Priya91** [comment]
@dotnet/dnceng There is no CI coverage for the PRs, all the Linux, windows, osx runs don't show the results of the test runs in mc.dot.net..

**Comment by Priya91** [comment]
> I didn't see any existing tests for the other verbs equivalent methods and properties so no unit tests are included.

We have tests for PutAsync, PostAsync, DeleteAsync and GetAsync in the Functional test project, you can do a string search. Please add a test for the newly added PatchAsync api.

**Comment by TylerBrinkley** [comment]
I added the functional tests. Sorry, I just used find references and didn't see any tests.

**Comment by TylerBrinkley** [comment]
@dotnet-bot test this please

**Comment by wfurt** [comment]
looks ok to me

**Comment by Priya91** [comment]
@dotnet-bot test outerloop Windows x64 debug build please
@dotnet-bot test outerloop Linux x64 debug build please
@dotnet-bot test outerloop osx x64 debug build please

**Comment by Priya91** [comment]
@TylerBrinkley the netfx test build failed with 

```
17:14:40 HttpClientTest.cs(272,34): error CS1061: 'HttpClient' does not contain a definition for 'PatchAsync' and no extension method 'PatchAsync' accepting a first argument of type 'HttpClient' could be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Net.Http\tests\FunctionalTests\System.Net.Http.Functional.Tests.csproj]
17:14:40 HttpClientTest.cs(305,37): error CS1061: 'HttpClient' does not contain a definition for 'PatchAsync' and no extension method 'PatchAsync' accepting a first argument of type 'HttpClient' could be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Net.Http\tests\FunctionalTests\System.Net.Http.Functional.Tests.csproj]
17:14:40 HttpClientTest.cs(306,37): error CS1061: 'HttpClient' does not contain a definition for 'PatchAsync' and no extension method 'PatchAsync' accepting a first argument of type 'HttpClient' could be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Net.Http\tests\FunctionalTests\System.Net.Http.Functional.Tests.csproj]
17:14:40 HttpClientTest.cs(377,67): error CS1061: 'HttpClient' does not contain a definition for 'PatchAsync' and no extension method 'PatchAsync' accepting a first argument of type 'HttpClient' could be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Net.Http\tests\FunctionalTests\System.Net.Http.Functional.Tests.csproj]
17:14:40 HttpMethodTest.cs(25,39): error CS0117: 'HttpMethod' does not contain a definition for 'Patch' [D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Net.Http\tests\FunctionalTests\System.Net.Http.Functional.Tests.csproj]
17:14:40 HttpMethodTest.cs(40,46): error CS0117: 'HttpMethod' does not contain a definition for 'Patch' [D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Net.Http\tests\FunctionalTests\System.Net.Http.Functional.Tests.csproj]
```

Since these are new apis only available in netcoreapp profile, these tests shouldn't be compiled or run for full framework. You should create a new test file ```HttpClientTest.netcoreapp.cs```, have a partial class with the ```patchasync``` tests and include it only for netcoreapp on csproj [here](https://github.com/TylerBrinkley/corefx/blob/d67255e150d3a12cabdc832c523be1c79db53e03/src/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj#L112)

**Comment by TylerBrinkley** [comment]
@Priya91 Thanks, I've updated the pr to make the patch method tests only apply to netcoreapp.

**Comment by TylerBrinkley** [comment]
@dotnet-bot test outerloop Windows x64 debug build please
@dotnet-bot test outerloop Linux x64 debug build please

**Comment by TylerBrinkley** [comment]
@dotnet-bot test Windows x64 debug build please
@dotnet-bot test Linux x64 debug build please

**Comment by TylerBrinkley** [comment]
@dotnet-bot test Linux x64 release build please

**Comment by TylerBrinkley** [comment]
@dotnet-bot test outerloop Linux x64 debug build please

File: src/System.Net.Http/tests/FunctionalTests/HttpClientTest.netcoreapp.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
```
Comment (Priya91) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientTest.netcoreapp.cs]: Missing license header..

File: src/System.Net.Http/tests/FunctionalTests/HttpMethodTest.netcoreapp.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Collections.Generic;
```
Comment (Priya91) [inline src/System.Net.Http/tests/FunctionalTests/HttpMethodTest.netcoreapp.cs]: missing license header..

**Comment by Priya91** [comment]
@TylerBrinkley Thanks for working on this, pending adding the license comments to the new files, changes LGTM. Will merge once you update that.

**Comment by JorgIanFloren** [comment]
Will it also be available in .NetStandard?

**Comment by davidsh** [comment]
>Will it also be available in .NetStandard?

These new API's cannot be in the current NETStandard2.0.  They could be in a future NETStandard. But in order for that to happen, these APIs must be ported and put into a future .NET Framework version.  Once in a future .NET Framework version, then a new NETStandard can be defined.

**Comment by JorgIanFloren** [comment]
Guess my solution for this stackoverflow thread won't be deprecated any time soon then

https://stackoverflow.com/questions/26218764/patch-async-requests-with-windows-web-http-httpclient-class/39450277#39450277

**Pull request merged**